src/Api.php

```diff
@@ -145,25 +166,179 @@ protected function computeHmac($hmac, $fields)
+     * @See documentation ManuelIntegrationVerifone_PayboxSystem_V8 5.3.4 Values checking
+     *
+     * @param GetHttpRequest $httpRequest
+     *
+     * @return bool
      */
-    public function getOptions()
+    public function verify($httpRequest)
     {
-        return $this->options;
+        $context = [
+            'httpRequestUri' => $httpRequest->uri,
+            'httpRequestQuery' => $httpRequest->query,
+        ];
+        if (false === $this->doesTheResponseReallyComeFromPaybox($httpRequest)) {
+            $message = 'Notification invalid: the response doesn\'t come from Paybox.';
+            $this->logger->error($message, $context);
+            throw new HttpResponse($message);
+        }
+
+        return true;
+    }
+
+    /**
+     * @param GetHttpRequest $httpRequest
+     *
+     * @return bool
+     */
+    private function doesTheResponseReallyComeFromPaybox($httpRequest)
+    {
+        // checks paybox signature
+        $this->logger->notice('checkPayboxSignature', [
+            'return' => $this->checkPayboxSignature($httpRequest),
+        ]);
+        if (false === $this->checkPayboxSignature($httpRequest)) {
+            return false;
+        }
+        // checks origin ip address
+        $this->logger->notice('checkOriginIpAddress', [
+            'return' => $this->checkOriginIpAddress($httpRequest),
+        ]);
+        if (false === $this->checkOriginIpAddress($httpRequest)) {
+            return false;
+        }
+
+        return true;
     }
 
+    /**
+     * @return bool
+     */
+    private function checkOriginIpAddress(GetHttpRequest $httpRequest)
+    {
+        return \in_array($httpRequest->clientIp, self::PAYBOX_IP_ADDRESSES, true);
+    }
+
+    /**
+     * @return bool
+     */
+    private function checkPayboxSignature(GetHttpRequest $httpRequest)
+    {
+        if (false === isset($httpRequest->query['signature'])) {
+            $this->logger->critical('checkPayboxSignature: query parameter \'signature\' not set', [
+                'query' => $httpRequest->query,
+            ]);
+
+            return false;
+        }
+
+        // init signature
+        $signature = $httpRequest->query['signature'];
+        $signatureLength = \strlen($signature);
+        if ($signatureLength > 172) {
+            $signatureInitialized = base64_decode(urldecode($signature));
+        } elseif (172 == $signatureLength) {
+            $signatureInitialized = base64_decode($signature);
+        } elseif (128 == $signatureLength) {
+            $signatureInitialized = $signature;
+        } else {
+            $signatureInitialized = null;
+            //throw new \Exception('Bad signature format.');
+        }
+        if (false === $signatureInitialized) {
+            $this->logger->error('Fail to base_decode signature', [
+                'signature' => $signature,
+            ]);
+
+            return false;
+        }
+
+        $file = fopen(__DIR__.'/Resources/pubkey.pem', 'rb');
```

Comment:
Where does this key come from ? Can we avoid to store this and check from an url instead ?

Reply:
The key comes from http://www1.paybox.com/espace-integrateur-documentation/manuels/?lang=en 
In the documentation it says 
> The
Paybox
public
key
is
available
for
download
on
the
website
http://www1.paybox.com/espace-integrateur-documentation/manuels/?lang=en. In order to
fulfill the security rules, Verifone may change his private/public key pair. Thus, it should be
possible to manage several keys in the same time in the merchant system.

i'm not favorable with an url.

Reply:
I understand, and we can't use an URL ! Let's keep that too for now ! We should think about a key repository as they mentioned that we may have more than one key.